### DISCUSSION TITLE
How is clear screen implemented?

### DISCUSSION THREAD

**Discussion by PatrickF1** [kovidgoyal/kitty#6674] [General]
Hi Kovid and other Kitty maintainers,

First off, thank you for making a terminal emulator that I love and has suited my needs almost perfectly! I am currently helping out on a new terminal emulator (https://mitchellh.com/ghostty). One thing we'd like to copy from Kitty is how it is able to clear the screen up to the prompt, but not the lines after the prompt if it's a multi line prompt when Ctrl+L has been bound within the shell to a function. I'm talking about this behavior:

https://github.com/kovidgoyal/kitty/assets/1967248/16f8ec89-ca06-49fd-afa0-cfe9090cd4fb

What I'm showing is that I bound Ctrl+L to ls, then typed a multi-line command, and then hit Cmd+K and it cleared the screen up to the prompt, but not below it. Our investigation so far has concluded that Kitty, on Cmd+K (at least on Mac) sends both the form-feed character (generated by Ctrl+L) and performs the clear_screen action. However, if it does that, how does Kitty not trigger the Ctrl+L key bind? When Ghostty copied that functionality, the screen clears but the Ctrl+L key bind also triggered.

https://github.com/kovidgoyal/kitty/assets/1967248/2b292fdc-7b5c-42c5-ac8c-2507d4a8ab06

What I'm showing here is that in Ghostty, I bound Ctrl+L to ls, and then hit Cmd+K. The screen cleared but ls was also triggered.
tl;dr how does Kitty implement the Cmd+K functionality to clear the screen up to the prompt but not below it, which when clearing the screen while typing a multi-line prompt?

**Comment by kovidgoyal** [comment]
https://sw.kovidgoyal.net/kitty/conf/#shortcut-kitty.Reset-the-terminal
